ml/src/main/scala/org/apache/spark/ml/h2o/models/H2OMOJOModel.scala

```diff
@@ -0,0 +1,222 @@
+  extends SparkModel[H2OMOJOModel] with H2OModelParams with MLWritable {
+
+  def this(model: MojoModel, mojoData: Array[Byte])(sqlContext: SQLContext) = this(model, mojoData, Identifiable.randomUID("mojoModel"))(sqlContext)
+
+  override def copy(extra: ParamMap): H2OMOJOModel = defaultCopy(extra)
+
+  private def flattenSchemaToCol(schema: StructType, prefix: String = null): Array[Column] = {
+    import org.apache.spark.sql.functions.col
+
+    schema.fields.flatMap { f =>
+      val colName = if (prefix == null) f.name else prefix + "." + f.name
+
+      f.dataType match {
+        case st: StructType => flattenSchemaToCol(st, colName)
+        case _ => Array[Column](col(colName))
+      }
+    }
+  }
+
+  def defaultFileName: String = H2OMOJOModel.defaultFileName
+
+  private def flattenDataFrame(df: DataFrame): DataFrame = {
+    import org.apache.spark.sql.functions.col
+    val flattenSchema = flattenSchemaToCol(df.schema)
+    // this is needed so the flattened data frame has hiearchical names
+    val renamedCols = flattenSchema.map(name => col(name.toString()).as(name.toString()))
+    df.select(renamedCols: _*)
+  }
+
+
+  override def transform(dataset: Dataset[_]): DataFrame = {
+
+
+    val spark = SparkSession.builder().getOrCreate()
+    val df = flattenDataFrame(dataset.toDF())
+
+    val predictedRows = df.rdd.map { row =>
+      // create RowData on which we do the predictions
+      val dt = new RowData
+      val values = row.schema.fields.zipWithIndex.map { case (entry, idxRow) =>
+
+        if ($(featuresCols).contains(entry.name)) {
+          setRowData(row, idxRow, dt, entry) // use only relevant columns for training
+        }
+
+        if (row.isNullAt(idxRow)) {
+          0
+        } else {
+          row.get(idxRow)
+        }
+      }
+      Row.merge(Row.fromSeq(values.toSeq), predict(dt))
+    }
+    spark.createDataFrame(predictedRows, StructType(df.schema.fields ++ getPredictionFrameSchema()))
+  }
+
+
+  def setRowData(row: Row, idxRow: Int, dt: RowData, entry: StructField) {
+    if (row.isNullAt(idxRow)) {
+      dt.put(entry.name, 0.toString) // 0 as NA
+    } else {
+      entry.dataType match {
+        case BooleanType =>
+          if (row.getBoolean(idxRow)) dt.put(entry.name, 1.toString) else dt.put(entry.name, 0.toString)
+        case BinaryType =>
+          row.getAs[Array[Byte]](idxRow).zipWithIndex.foreach { case (v, idx) =>
+            dt.put(entry.name + idx, v.toString)
+          }
+        case ByteType => dt.put(entry.name, row.getByte(idxRow).toString)
+        case ShortType => dt.put(entry.name, row.getShort(idxRow).toString)
+        case IntegerType => dt.put(entry.name, row.getInt(idxRow).toString)
+        case LongType => dt.put(entry.name, row.getLong(idxRow).toString)
+        case FloatType => dt.put(entry.name, row.getFloat(idxRow).toString)
+        case _: DecimalType => dt.put(entry.name, row.getDecimal(idxRow).doubleValue().toString)
```

Comment:
Not sure at the moment, maybe it's not necessary. However we need this if the user is allowed to create objects of type DecimalType.

 I would just firstly check if all the types above fill the domain of decimal type ( i.e. That we cover all decimal) and that the DecimalType can't be instantiated. Then we can remove it, otherwise we need to leave it there.